src/ctap/token_state.rs

```diff
@@ -0,0 +1,277 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::ctap::client_pin::PinPermission;
+use crate::ctap::status_code::Ctap2StatusCode;
+use crate::ctap::timed_permission::TimedPermission;
+use alloc::string::String;
+use crypto::sha256::Sha256;
+use crypto::Hash256;
+use libtock_drivers::timer::{ClockValue, Duration};
+
+/// Timeout for auth tokens.
+///
+/// This usage time limit is correct for USB, BLE, and internal.
+/// NFC only allows 19.8 seconds.
+/// TODO(#15) multiplex over transports, add NFC
+const INITIAL_USAGE_TIME_LIMIT: Duration<isize> = Duration::from_ms(30000);
+
+/// Implements pinUvAuthToken state from section 6.5.2.1.
+///
+/// The userPresent flag is omitted as the only way to set it to true is
+/// built-in user verification. Therefore, we never cache user presence.
+///
+/// This implementation does not use a rolling timer.
+pub struct PinUvAuthTokenState {
+    // Relies on the fact that all permissions are represented by powers of two.
+    permissions_set: u8,
```

Comment:
If this is a bitfield, I wonder if we couldn't reuse some of the boilerplate that exists in Tock for the registers

Reply:
In the specification, `permissions` is an "Unsigned Integer" described as
> Bitfield of permissions. If present, MUST NOT be 0. See § 6.5.5.7 Operations to Obtain a pinUvAuthToken.

This implementation tries to hide the bitfield as an implementation detail, allowing read access only through the `PinPermission` enum. Therefore, I went with the easiest and most direct implementation. Is this the crate you talk about?
[https://crates.io/crates/tock-registers](https://crates.io/crates/tock-registers)
I'd be worried about mixing register semantics in.

Reply:
Yes, this crate.
But that was just food for thoughts because it's made to deal with bitfields in a nice way. Nothing blocking the merge here or I wouldn't have approved the PR ;-)

Reply:
Alright, thanks!